### PR TITLE
feat: add tcp packet latency sampler

### PR DESCRIPTION
**Pull request by brayniac**
Adds a TCP packet latency sampler that measures the latency from as packet being received to being handled by userspace application.

### CODE REVIEW

File: src/samplers/tcp/packet_latency/mod.bpf.c [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+	if (targ_sport && targ_sport != BPF_CORE_READ(inet, inet_sport))
+		return 0;
+	if (targ_dport && targ_dport != BPF_CORE_READ(sk, __sk_common.skc_dport))
+		return 0;
+	th = (const struct tcphdr*)BPF_CORE_READ(skb, data);
+	doff = BPF_CORE_READ_BITFIELD_PROBED(th, doff);
+	len = BPF_CORE_READ(skb, len);
+	/* `doff * 4` means `__tcp_hdrlen` */
+	if (len <= doff * 4)
```
Comment (thinkingfish) [inline src/samplers/tcp/packet_latency/mod.bpf.c:62]: let's wrap this in `{}` just for good measure.

File: src/samplers/tcp/packet_latency/mod.bpf.c [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	u64 sock_ident = get_sock_ident(sk);
+	u64 id = bpf_get_current_pid_tgid(), *tsp;
+	u32 idx;
+	u64 now, delta_ns, *cnt;
+	u32 pid = id >> 32, tid = id;
+	struct event *eventp;
+	u16 family;
+
+	tsp = bpf_map_lookup_elem(&start, &sock_ident);
+	if (!tsp)
```
Comment (thinkingfish) [inline src/samplers/tcp/packet_latency/mod.bpf.c:82]: ditto

File: src/samplers/tcp/packet_latency/mod.bpf.c [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		return 0;
+
+	now = bpf_ktime_get_ns();
+
+	if (*tsp > now)
```
Comment (thinkingfish) [inline src/samplers/tcp/packet_latency/mod.bpf.c:87]: ditto

**Review by thinkingfish — COMMENTED** [review]
What happens when multiple packets are received before a user space copy is called? Would the timestamp in the map be overridden each time?

**Comment by brayniac** [comment]
> What happens when multiple packets are received before a user space copy is called? Would the `start` timestamp in the map be overridden each time?

Yes, that appears to be correct. Should probably change the behavior there.

**Pull request merged**